[NVFuser] make comparators obey strict weak ordering

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75766
* __->__ #75713

Internally there is an at-runtime assert that was failing. The assert is checking to make sure that comparators obey strict weak ordering
https://en.cppreference.com/w/cpp/named_req/Compare

In particular `comp(a, a) == false` was failing on lower_loops.cpp.

Differential Revision: [D35599653](https://our.internmc.facebook.com/intern/diff/D35599653)

davidberard98 committed:
[NVFuser] make comparators obey strict weak ordering

Internally there is an at-runtime assert that was failing. The assert is checking to make sure that comparators obey strict weak ordering
https://en.cppreference.com/w/cpp/named_req/Compare

In particular `comp(a, a) == false` was failing.

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35599653).

davidberard98 committed:
Update on "[NVFuser] make comparators obey strict weak ordering"


Internally there is an at-runtime assert that was failing. The assert is checking to make sure that comparators obey strict weak ordering
https://en.cppreference.com/w/cpp/named_req/Compare

In particular `comp(a, a) == false` was failing.

Differential Revision: [D35599653](https://our.internmc.facebook.com/intern/diff/D35599653)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35599653).

jjsjann123 commented on torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp:
```diff
@@ -691,17 +691,21 @@ struct LocalDomainSorter {
       const auto& dependencies_1 = concrete_id_dependencies_.at(concrete_id_1);
       // if id1 depends on id0 it means id0 is inside id1, so id1 < id0
-      return !dependencies_1.count(concrete_id_0);
+      if (dependencies_1.count(concrete_id_0)) {
```
This doesn't look right... We can only return `false` after this point. In the old code, we could return `true` if `dependencies_1.count(concrete_id_0) == false`

davidberard98 replied:
this is my understanding, but lmk if something seems wrong here:

- this looks like it's basically trying to do a toposort, where, if a depends on b, then we want a to go before b in the final sorted order. afaik, the typical toposort comparator is "if a path a->b exists, then comp(a, b) = true. Otherwise comp(a, b) = false". Or in this case, if a maps to a set containing b, then return comp(a, b) = true. In particular, if there's no relationship between a and b, then comp(a, b) and comp(b, a) should both be false (unless you have some other way to break ties in this case)

if this is a toposort, then `dependencies_1.count(concrete_id_0) == false` is a case where  we don't necessarily have any dependence between `dep_1` and `dep_0`; in other words, `0 -> 1` (dep_0 maps to dep_1) implies we should return true; and `0 -> 1` implies `1 -/-> 0`; but `1 -/-> 0` doesn't imply `0 -> 1`, so `1 -/-> 0` doesn't necessarily mean we should return true.

jjsjann123 replied:
This sounds very reasonable to me (without knowing much about the actual implementation).

I'd argue in that instance, the entire block inside could just be removed, since it only returns `false` inside and that is the default behavior after the true block. https://github.com/pytorch/pytorch/blob/9b49a26918f1a91daf34233036eddb90d1a2ae71/torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp#L699-L706

Having said that, I'm waiting on someone who actually knows what this does to respond. cc'ing @csarofeen @naoyam

davidberard98 replied:
yeah, that's true - I can delete those blocks if we decide this change is okay

naoyam replied:
Thanks @davidberard98. I believe the fix makes sense and is indeed a nice catch.

davidberard98 replied:
great, thanks for the help reviewing @naoyam @jjsjann123 

I removed the false blocks like @jjsjann123 suggested

jjsjann123 commented on torch/csrc/jit/codegen/cuda/lower_loops.cpp:
```diff
@@ -167,17 +167,21 @@ struct LocalDomainSorter {
       const auto& dependencies_1 = concrete_id_dependencies_.at(concrete_id_1);
       // if id1 depends on id0 it means id1 is inside id0, so id0 < id1
-      return dependencies_1.count(concrete_id_0);
+      if (dependencies_1.count(concrete_id_0)) {
```
ditto

jjsjann123 replied:
IIUC, the first block and the second block are checking the same thing from different perspective and that's what strict weak ordering checks as well. So even though it looks slightly strange from the code diff, they are still correct.

I'm stamping it.

jjsjann123 commented:
Fort this check `comp(a, a) == false`

Maybe we can add a quick short cut where we just return `false` when both things are pointing to the same value?! cc'ing @csarofeen

davidberard98 committed:
Update on "[NVFuser] make comparators obey strict weak ordering"


Internally there is an at-runtime assert that was failing. The assert is checking to make sure that comparators obey strict weak ordering
https://en.cppreference.com/w/cpp/named_req/Compare

In particular `comp(a, a) == false` was failing on lower_loops.cpp.

Differential Revision: [D35599653](https://our.internmc.facebook.com/intern/diff/D35599653)

[ghstack-poisoned]

jjsjann123 reviewed (approved):
LGTM

csarofeen commented:
Thanks @davidberard98 let me check this quickly today and will get back to you on the changes.

csarofeen commented:
Thanks again, @davidberard98. I was just trying to confirm some properties of this sorting. So what they're actually doing is a "topo sort" of "loop nests". I naively assumed when I wrote that code that there was a single unique ordering of loop nests, but in fact it makes sense that in certain kernels that's not the case, and multiple orderings can be correct. As long as both orderings consistent (perfectly reversed actually), then we should be fine. concrete_id_1 and concrete_id_2 in both of these blocks should have the same exact information, we just want sorting order to be exactly reversed.

So the block in lower_expr_sort the old logic was:

```
  if(id0 dependent on id1)
    id0 < id1
  else if(id1 dependent on id0)
    id0 >= id1
  else
    id0 < id1
```

and in lower_loops:

```
  if(id0 dependent on id1)
    id0 >= id1
  else if(id1 dependent on id0)
    id0 < id1
  else
    id0 < id1
```

With your updated logic I believe it's in lower_expr_sort:
```
  if(id0 dependent on id1)
    id0 < id1
  else
    id0 >= id1
```

and in lower_loops:
```
  if(id0 dependent on id1)
    id0 >= id1
  else
    id0 < id1
```

So a few things. I agree there's an issue with unstable sorting. What I think we should do is in lower_loops put exactly what's in lower_expr_sort and reverse the sorted list used in lower_loops.cpp. Does that make sense @davidberard98 and @naoyam?

naoyam commented:
Reversing the order generated in lower_expr_sort seems better as it removes the logic duplication.

Re:
```
and in lower_loops:

  if(id0 dependent on id1)
    id0 >= id1
  else
    id0 < id1
```

I think what the PR does is to check if id1 depends on id0, and only in that case returns true.

davidberard98 committed:
Update on "[NVFuser] make comparators obey strict weak ordering"


Internally there is an at-runtime assert that was failing. The assert is checking to make sure that comparators obey strict weak ordering
https://en.cppreference.com/w/cpp/named_req/Compare

In particular `comp(a, a) == false` was failing on lower_loops.cpp.

Differential Revision: [D35599653](https://our.internmc.facebook.com/intern/diff/D35599653)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35599653).

naoyam commented on torch/csrc/jit/codegen/cuda/lower_utils.h:
```diff
@@ -140,6 +141,62 @@ std::vector<Expr*> replaceInputsInExpr(
+
+  const std::unordered_map<IterDomain*, std::unordered_set<IterDomain*>>&
+      concrete_id_dependencies_;
+  const ComputeAtMap compute_at_map_;
```
nit: this can be a const reference.

davidberard98 replied:
whoops, meant to do that - thanks for noticing this!

naoyam commented:
Thanks again @davidberard98 . The update looks good to me. Just added a minor comment.

davidberard98 committed:
Update on "[NVFuser] make comparators obey strict weak ordering"


Internally there is an at-runtime assert that was failing. The assert is checking to make sure that comparators obey strict weak ordering
https://en.cppreference.com/w/cpp/named_req/Compare

In particular `comp(a, a) == false` was failing on lower_loops.cpp.

Differential Revision: [D35599653](https://our.internmc.facebook.com/intern/diff/D35599653)

[ghstack-poisoned]

csarofeen reviewed (approved):
Thanks so much for the help!

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35599653).

davidberard98 commented:
@pytorchmergebot merge this
